README.md
![Build](https://github.com/sarmkadan/signalr-map-realtime/actions/workflows/build.yml/badge.svg)
![License](https://img.shields.io/github/license/sarmkadan/signalr-map-realtime)
![.NET](https://img.shields.io/badge/.NET-10.0-512BD4)

# SignalR Map Realtime - Real-Time Location Tracking System

A production-grade, enterprise-ready real-time location tracking system built with ASP.NET Core 10, SignalR, and modern web technologies. Track vehicles, couriers, assets, and mobile workers in real-time with live GPS updates, intelligent route management, and comprehensive fleet monitoring.

## Table of Contents

- [Overview](#overview)
- [Key Features](#key-features)
- [Technical Stack](#technical-stack)
- [Architecture](#architecture)
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Usage Examples](#usage-examples)
- [API Reference](#api-reference)
- [Configuration Guide](#configuration-guide)
- [Self-Hosting Guide](#self-hosting-guide)
- [Advanced Topics](#advanced-topics)
- [Performance](#performance)
- [Troubleshooting](#troubleshooting)
- [Testing](#testing)
- [Related Projects](#related-projects)
- [Contributing](#contributing)
- [License](#license)

## Overview

SignalR Map Realtime provides a comprehensive solution for tracking the location of mobile assets in real-time. Whether you're building a ride-sharing platform, fleet management system, logistics solution, or any application requiring live location tracking, this project provides a battle-tested, scalable foundation.

The system leverages WebSocket technology through SignalR for ultra-low-latency bidirectional communication, enabling seamless real-time updates of vehicle positions, route progress, and system alerts. The architecture follows domain-driven design principles, ensuring clean separation of concerns and long-term maintainability.

### Use Cases

- **Fleet Management**: Monitor vehicle locations, fuel consumption, driver behavior
- **Ride-Sharing**: Real-time driver availability, ETA calculation, dynamic routing
- **Logistics & Delivery**: Package tracking, route optimization, delivery proof
- **Field Service**: Technician location, job assignment, service area coverage
- **Public Transportation**: Bus/vehicle tracking, arrival predictions, crowd management
- **Emergency Response**: Ambulance/fire truck dispatching, optimal routing

## Key Features

### Core Tracking Capabilities

- **Real-time GPS Updates**: Bi-directional WebSocket communication via SignalR for instantaneous location streaming
- **Multi-Asset Tracking**: Simultaneous tracking of vehicles, couriers, packages, and equipment
- **Historical Tracking**: Complete location history with timestamps and metadata
- **Session Management**: Start, pause, resume, and complete tracking sessions with statistics
- **Location Clustering**: Efficiently group nearby locations to reduce bandwidth

### Advanced Features

- **Route Planning & Management**: Create optimized routes with multiple waypoints
- **Geofencing**: Define geographic boundaries with entry/exit notifications
- **Live Statistics**: Real-time distance, speed, and performance calculations
- **Vehicle Status Tracking**: Online/offline status, connectivity monitoring
- **Driver/Courier Management**: User assignments and performance tracking
- **Asset Type Support**: Vehicles, packages, equipment with type-specific attributes
- **Speed Monitoring**: Detect and alert on speed violations
- **Fuel Tracking**: Monitor fuel levels and consumption patterns

### Technical Features

- **Comprehensive API**: RESTful endpoints with full Swagger/OpenAPI documentation
- **Authentication Ready**: API key authentication infrastructure for extensibility
- **Rate Limiting**: Built-in rate limiting to prevent abuse
- **Caching Layer**: Redis-compatible in-memory caching for performance
- **Error Handling**: Comprehensive error handling with custom exceptions
- **Request Logging**: Detailed request/response logging for debugging
- **Performance Monitoring**: Middleware for tracking endpoint performance
- **Webhook Support**: Integration points for third-party systems

## Technical Stack

| Component | Technology | Version |
|-----------|-----------|---------|
| **Runtime** | .NET | 10.0 |
| **Framework** | ASP.NET Core | 10.0 |
| **Language** | C# | 13.0 |
| **ORM** | Entity Framework Core | 10.0 |
| **Real-time** | SignalR | 10.0 |
| **Database** | SQL Server / PostgreSQL | 2019+ |
| **Object Mapping** | AutoMapper | 13.0 |
| **Validation** | FluentValidation | 11.9+ |
| **API Docs** | Swagger/OpenAPI | 3.0 |
| **Caching** | In-Memory/Redis | - |
| **Frontend** | Leaflet.js | 1.9+ |

## Architecture

### Component Interaction

The three main layers — Leaflet frontend, ASP.NET Core backend, and the database — communicate as follows:

```
  Browser (Leaflet.js)
        │
        │  1. HTTP REST  ──────────────────────────────────────────────────►
        │                                                  ASP.NET Core App
        │  2. WebSocket (SignalR)  ────────────────────────────────────────►
        │                                                  LocationHub
        │                                                       │
        │                                               LocationService
        │                                               VehicleService
        │                                                       │
        │                                               EF Core / SQL Server
        │                                                       │
        │  3. SignalR broadcast ◄──────────────────────────────────────────
        │     "LocationUpdated",
        │     "PositionSnapshot",
        │     "VehicleStatusChanged", …
        │
  Leaflet map updates marker positions
```

**Data flow for a single location update:**

1. A tracked device calls `POST /api/v1/locations` or invokes `SendLocationUpdate` on the `LocationHub`.
2. `LocationService` validates coordinates, persists the record, and updates the vehicle's last-known position.
3. `LocationHub` broadcasts `LocationUpdated` to **all** connected clients and `VehicleLocationUpdated` to the vehicle-specific group.
4. Each browser's SignalR listener calls `updateVehicleLocation()`, which updates the Leaflet marker position.

**Reconnect handling:**

When a client's WebSocket drops and re-establishes, the `onreconnected` callback invokes `RequestAllVehicleLocations` on the hub. The hub responds with a `PositionSnapshot` containing the latest known position of every online vehicle, so all Leaflet markers are immediately resynced without waiting for the next delta update.

### Infrastructure Requirements

| Concern | Single-node | Multi-node (scale-out) |
|---------|-------------|------------------------|
| **Transport** | WebSockets (default) | WebSockets; sticky sessions **required** unless a backplane is configured |
| **SignalR backplane** | Not required | Redis or Azure SignalR Service |
| **Session affinity** | Not required | Required when backplane is absent (configure on load balancer) |
| **Database** | SQL Server (LocalDB for dev) | SQL Server with connection pooling |
| **Cache** | In-memory | Redis |

#### Enabling the Redis backplane (multi-node)

```bash
dotnet add package Microsoft.AspNetCore.SignalR.StackExchangeRedis
```

```json
// appsettings.json
{
  "SignalR": {
    "Redis": {
      "ConnectionString": "localhost:6379"
    }
  }
}
```

```csharp
// Program.cs
builder.Services.AddSignalR()
    .AddStackExchangeRedis(builder.Configuration["SignalR:Redis:ConnectionString"]!);
```

### System Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                       Web Clients                               │
│              (Browser, Mobile, Desktop Apps)                   │
└────────────────┬────────────────────────────────────────────────┘
                 │
         ┌───────┴────────┐
         │                │
    ┌────▼──────┐    ┌───▼─────┐
    │   HTTP    │    │WebSocket │
    │   (REST)  │    │(SignalR) │
    └────┬──────┘    └───┬─────┘
         │                │
    ┌────▼────────────────▼──────────────────────┐
    │         ASP.NET Core Application           │
    │  ┌─────────────────────────────────────┐   │
    │  │    API Controllers (REST)           │   │
    │  │    SignalR Hubs (Real-time)         │   │
    │  │    Middleware Pipeline              │   │
    │  │    ├─ Authentication                │   │
    │  │    ├─ CORS                          │   │
    │  │    ├─ Rate Limiting                 │   │
    │  │    ├─ Error Handling                │   │
    │  │    └─ Logging                       │   │
    │  └────────────────┬──────────────────┘   │
    │                   │                      │
    │  ┌────────────────▼──────────────────┐   │
    │  │      Business Service Layer       │   │
    │  │  ├─ LocationService               │   │
    │  │  ├─ VehicleService                │   │
    │  │  ├─ TrackingService               │   │
    │  │  ├─ NotificationService           │   │
    │  │  └─ CacheService                  │   │
    │  └────────────────┬──────────────────┘   │
    │                   │                      │
    │  ┌────────────────▼──────────────────┐   │
    │  │    Data Access Layer              │   │
    │  │  (Repository Pattern)             │   │
    │  │  ├─ LocationRepository            │   │
    │  │  ├─ VehicleRepository             │   │
    │  │  ├─ RouteRepository               │   │
    │  │  ├─ AssetRepository               │   │
    │  │  └─ UserRepository                │   │
    │  └────────────────┬──────────────────┘   │
    │                   │                      │
    │  ┌────────────────▼──────────────────┐   │
    │  │  Entity Framework Core (EF Core)  │   │
    │  │  ├─ DbContext                     │   │
    │  │  ├─ Migrations                    │   │
    │  │  └─ LINQ to Entities              │   │
    │  └────────────────┬──────────────────┘   │
    └────────────────┬──────────────────────────┘
                     │
    ┌────────────────▼──────────────────┐
    │      SQL Server Database          │
    │  ├─ Vehicles                      │
    │  ├─ Locations (History)           │
    │  ├─ Routes & Waypoints            │
    │  ├─ Users (Drivers/Couriers)      │
    │  ├─ Assets                        │
    │  ├─ TrackingSessions              │
    │  └─ Indices & Views               │
    └───────────────────────────────────┘
```

### Layered Architecture

```
┌─────────────────────────────────┐
│    Presentation Layer           │
│  (Controllers, SignalR Hubs)    │
└────────────────┬────────────────┘
                 │
┌────────────────▼────────────────┐
│     Application/Service Layer   │
│    (Business Logic, DTOs)       │
└────────────────┬────────────────┘
                 │
┌────────────────▼────────────────┐
│     Domain Layer                │
│   (Models, Enums, Exceptions)   │
└────────────────┬────────────────┘
                 │
┌────────────────▼────────────────┐
│    Data Access Layer            │
│   (Repositories, EF Core)       │
└────────────────┬────────────────┘
                 │
┌────────────────▼────────────────┐
│     Infrastructure Layer        │
│   (Database, Configuration)     │
└─────────────────────────────────┘
```

### Project Structure

```
signalr-map-realtime/
├── src/
│   └── SignalRMapRealtime/
│       ├── Domain/
│       │   ├── Models/
│       │   │   ├── Vehicle.cs
│       │   │   ├── Location.cs
│       │   │   ├── Route.cs
│       │   │   ├── User.cs
│       │   │   ├── Asset.cs
│       │   │   ├── Waypoint.cs
│       │   │   └── TrackingSession.cs
│       │   └── Enums/
│       │       ├── VehicleStatus.cs
│       │       ├── LocationType.cs
│       │       ├── AssetType.cs
│       │       └── SessionStatus.cs
│       ├── Data/
│       │   ├── ApplicationDbContext.cs
│       │   └── Repositories/
│       │       ├── IRepository.cs
│       │       ├── BaseRepository.cs
│       │       ├── VehicleRepository.cs
│       │       ├── LocationRepository.cs
│       │       ├── RouteRepository.cs
│       │       ├── AssetRepository.cs
│       │       ├── UserRepository.cs
│       │       └── TrackingSessionRepository.cs
│       ├── Services/
│       │   ├── LocationService.cs
│       │   ├── VehicleService.cs
│       │   ├── TrackingService.cs
│       │   ├── NotificationService.cs
│       │   ├── CacheService.cs
│       │   └── Interfaces/
│       │       ├── ILocationService.cs
│       │       ├── IVehicleService.cs
│       │       └── ITrackingService.cs
│       ├── Controllers/
│       │   ├── VehicleController.cs
│       │   ├── LocationController.cs
│       │   ├── RouteController.cs
│       │   └── AssetController.cs
│       ├── Hubs/
│       │   └── LocationHub.cs
│       ├── DTOs/
│       │   ├── VehicleDto.cs
│       │   ├── LocationDto.cs
│       │   ├── RouteDto.cs
│       │   ├── AssetDto.cs
│       │   └── UserDto.cs
│       ├── Configuration/
│       │   ├── DependencyInjection.cs
│       │   ├── AutoMapperProfile.cs
│       │   └── Options/
│       ├── Middleware/
│       │   ├── ErrorHandlingMiddleware.cs
│       │   ├── LoggingMiddleware.cs
│       │   ├── RateLimitingMiddleware.cs
│       │   └── PerformanceMonitoringMiddleware.cs
│       ├── Attributes/
│       │   └── ApiKeyAuthenticationAttribute.cs
│       ├── Exceptions/
│       │   └── LocationTrackingException.cs
│       ├── BackgroundJobs/
│       │   └── SessionCleanupWorker.cs
│       ├── Events/
│       │   ├── DomainEvent.cs
│       │   └── EventBus.cs
│       ├── Integration/
│       │   ├── HttpClientFactory.cs
│       │   └── WebhookHandler.cs
│       ├── Utilities/
│       │   └── Extension methods
│       ├── Constants/
│       ├── Program.cs
│       ├── appsettings.json
│       ├── appsettings.Development.json
│       └── SignalRMapRealtime.csproj
├── examples/
│   ├── VehicleTrackerClient.cs
│   ├── LocationUpdateSimulator.cs
│   ├── RouteOptimizationClient.cs
│   ├── FleetManagementDashboard.cs
│   ├── GeofenceAlertsClient.cs
│   ├── BulkLocationImporter.cs
│   ├── SessionAnalyticsReporter.cs
│   └── WebSocketRealTimeViewer.html
├── docs/
│   ├── getting-started.md
│   ├── architecture.md
│   ├── api-reference.md
│   ├── deployment.md
│   ├── faq.md
│   ├── database-schema.md
│   └── troubleshooting.md
├── .github/
│   └── workflows/
│       └── build.yml
├── Dockerfile
├── docker-compose.yml
├── .editorconfig
├── Makefile
├── CHANGELOG.md
├── LICENSE
├── .gitignore
└── signalr-map-realtime.sln
```

## Installation

### Prerequisites

- **.NET SDK**: 10.0 or later ([download](https://dotnet.microsoft.com/download/dotnet/10.0))
- **SQL Server**: 2019 Express or later, or use LocalDB
- **Visual Studio**: 2022 (17.8+) or VS Code with C# extension
- **Git**: For cloning the repository

### Method 1: Using Visual Studio (Recommended for Windows)

1. **Clone the Repository**
   ```bash
   git clone https://github.com/Sarmkadan/signalr-map-realtime.git
   cd signalr-map-realtime
   ```

2. **Open Solution**
   - Open `signalr-map-realtime.sln` in Visual Studio 2022

3. **Configure Connection String** (if needed)
   - Edit `src/SignalRMapRealtime/appsettings.json`
   - Update `ConnectionStrings.DefaultConnection` for your SQL Server

4. **Apply Migrations**
   - Package Manager Console: `Update-Database`
   - Or CLI: `dotnet ef database update`

5. **Run the Application**
   - Press `F5` or click Run button
   - API available at `https://localhost:5001`
   - Swagger UI at `https://localhost:5001/swagger`

### Method 2: Command Line

```bash
# Clone
git clone https://github.com/Sarmkadan/signalr-map-realtime.git
cd signalr-map-realtime

# Restore packages
dotnet restore

# Build
dotnet build

# Apply database migrations
dotnet ef database update --project src/SignalRMapRealtime

# Run
dotnet run --project src/SignalRMapRealtime
```

### Method 3: Docker

```bash
# Clone and navigate
git clone https://github.com/Sarmkadan/signalr-map-realtime.git
cd signalr-map-realtime

# Build image
docker build -t signalr-map-realtime:latest .

# Run with compose (includes SQL Server)
docker-compose up -d

# API available at http://localhost:5000
# Swagger at http://localhost:5000/swagger
```

## Quick Start

### 1. Verify Installation

```bash
curl https://localhost:5001/health
# Expected response: { "status": "Healthy" }
```

### 2. Create a Vehicle

```bash
curl -X POST https://localhost:5001/api/v1/vehicles \
  -H "Content-Type: application/json" \
  -d '{
    "licensePlate": "ABC-123",
    "manufacturer": "Toyota",
    "model": "Camry",
    "year": 2023,
    "status": "Available"
  }'
```

### 3. Start Real-time Tracking

```javascript
// Connect to SignalR hub
const connection = new signalR.HubConnectionBuilder()
    .withUrl("https://localhost:5001/locationHub")
    .withAutomaticReconnect()
    .build();

// Subscribe to location updates
connection.on("LocationUpdated", (location) => {
    console.log(`Vehicle ${location.vehicleId} at ${location.latitude}, ${location.longitude}`);
});

// Connect and send location
await connection.start();
await connection.invoke("SendLocationUpdate", {
    vehicleId: "your-vehicle-id",
    latitude: 40.7128,
    longitude: -74.0060,
    accuracy: 5.0
});
```

### 4. Retrieve Location History

```bash
curl "https://localhost:5001/api/v1/locations/vehicle/your-vehicle-id" \
  -H "Authorization: Bearer YOUR_TOKEN"
```

## Usage Examples

### Example 1: Basic Vehicle Tracking

See `examples/VehicleTrackerClient.cs` for a complete working example.

```csharp
var client = new HttpClient();
var baseUrl = "https://localhost:5001";

// Create vehicle
var vehicle = new { 
    licensePlate = "XYZ-789",
    manufacturer = "Ford",
    model = "Transit"
};

var response = await client.PostAsJsonAsync(
    $"{baseUrl}/api/v1/vehicles", 
    vehicle
);
var vehicleId = await response.Content.ReadAsAsync<string>();

// Start tracking
var location = new {
    vehicleId = vehicleId,
    latitude = 40.7128,
    longitude = -74.0060,
    accuracy = 10.0,
    speed = 35.5,
    heading = 90,
    timestamp = DateTime.UtcNow
};

await client.PostAsJsonAsync(
    $"{baseUrl}/api/v1/locations",
    location
);
```

### Example 2: SignalR Real-time Updates

See `examples/WebSocketRealTimeViewer.html` for a complete HTML/JavaScript example.

```csharp
// In your C# client
var connection = new HubConnectionBuilder()
    .WithUrl("https://localhost:5001/locationHub")
    .WithAutomaticReconnect()
    .Build();

connection.On<LocationDto>("LocationUpdated", location => 
{
    Console.WriteLine($"Update: {location.VehicleId} - {location.Latitude}, {location.Longitude}");
});

connection.On<string, string>("VehicleStatusChanged", (vehicleId, status) =>
{
    Console.WriteLine($"Vehicle {vehicleId} is now {status}");
});

await connection.StartAsync();

// Subscribe to specific vehicle
await connection.InvokeAsync("SubscribeToVehicle", vehicleId);
```

### Example 3: Route Optimization

See `examples/RouteOptimizationClient.cs`.

```csharp
var route = new {
    name = "Downtown Delivery Route",
    vehicleId = "vehicle-123",
    waypoints = new[] {
        new { latitude = 40.7128, longitude = -74.0060, name = "Start" },
        new { latitude = 40.7489, longitude = -73.9680, name = "Stop 1" },
        new { latitude = 40.7614, longitude = -73.9776, name = "Stop 2" },
        new { latitude = 40.7282, longitude = -73.7949, name = "End" }
    },
    estimatedDuration = 3600
};

var response = await client.PostAsJsonAsync(
    $"{baseUrl}/api/v1/routes",
    route
);
```

### Example 4: Geofence Alerts

See `examples/GeofenceAlertsClient.cs`.

```csharp
connection.On<AlertDto>("Alert", alert =>
{
    if (alert.Type == AlertType.GeofenceEntry)
    {
        Console.WriteLine($"Vehicle {alert.VehicleId} entered restricted area!");
    }
    else if (alert.Type == AlertType.GeofenceExit)
    {
        Console.WriteLine($"Vehicle {alert.VehicleId} left geofence");
    }
});
```

### Example 5: Bulk Location Import

See `examples/BulkLocationImporter.cs` for importing historical data.

```csharp
var locations = new List<LocationDto>();
// ... populate locations ...

var client = new HttpClient();
foreach (var batch in locations.Batch(100))
{
    await client.PostAsJsonAsync($"{baseUrl}/api/v1/locations/bulk", batch);
}
```

### Example 6: Session Analytics

See `examples/SessionAnalyticsReporter.cs`.

```csharp
var session = await trackingService.GetSessionAsync(sessionId);
var analytics = new {
    sessionId = session.Id,
    totalDistance = session.TotalDistance,
    totalDuration = session.EndTime - session.StartTime,
    averageSpeed = session.AverageSpeed,
    maxSpeed = session.MaxSpeed,
    locationCount = session.Locations.Count
};
```

## API Reference

### Health & Info Endpoints

#### Health Check
```
GET /health
```
**Response** (200 OK):
```json
{
  "status": "Healthy",
  "timestamp": "2024-05-04T10:30:00Z",
  "version": "2.0.2"
}
```

#### API Information
```
GET /api/info
```
**Response** (200 OK):
```json
{
  "version": "2.0.2",
  "title": "SignalR Map Realtime API",
  "environment": "Production"
}
```

### Vehicle Endpoints

#### List Vehicles
```
GET /api/v1/vehicles
?skip=0&take=20&status=Active
```
**Query Parameters**:
- `skip`: Number of items to skip (default: 0)
- `take`: Number of items to return (default: 20, max: 100)
- `status`: Filter by status (Active, Inactive, InMaintenance)

**Response** (200 OK):
```json
{
  "data": [
    {
      "id": "uuid",
      "licensePlate": "ABC-123",
      "manufacturer": "Toyota",
      "model": "Camry",
      "year": 2023,
      "status": "Active",
      "lastLocationUpdate": "2024-05-04T10:25:00Z",
      "currentLatitude": 40.7128,
      "currentLongitude": -74.0060,
      "speed": 35.5
    }
  ],
  "pageSize": 20,
  "totalCount": 150
}
```

#### Get Vehicle Details
```
GET /api/v1/vehicles/{id}
```
**Response** (200 OK):
```json
{
  "id": "uuid",
  "licensePlate": "ABC-123",
  "manufacturer": "Toyota",
  "model": "Camry",
  "year": 2023,
  "status": "Active",
  "fuelLevel": 85.5,
  "odometerReading": 45230,
  "assignedDriver": {
    "id": "uuid",
    "name": "John Doe",
    "phone": "+1234567890"
  },
  "createdAt": "2024-01-15T09:00:00Z",
  "updatedAt": "2024-05-04T10:25:00Z"
}
```

#### Create Vehicle
```
POST /api/v1/vehicles
Content-Type: application/json

{
  "licensePlate": "XYZ-789",
  "manufacturer": "Ford",
  "model": "Transit",
  "year": 2024,
  "vin": "1FTYR14D04TM00001",
  "status": "Active"
}
```
**Response** (201 Created):
```json
{
  "id": "uuid",
  "licensePlate": "XYZ-789",
  "status": "Active"
}
```

#### Update Vehicle
```
PUT /api/v1/vehicles/{id}
Content-Type: application/json

{
  "licensePlate": "XYZ-789",
  "status": "InMaintenance",
  "fuelLevel": 75.0
}
```
**Response** (204 No Content)

#### Delete Vehicle
```
DELETE /api/v1/vehicles/{id}
```
**Response** (204 No Content)

### Location Endpoints

#### Get Vehicle Locations
```
GET /api/v1/locations/vehicle/{vehicleId}
?startDate=2024-05-01&endDate=2024-05-04&limit=1000
```
**Query Parameters**:
- `startDate`: Filter from date (ISO 8601)
- `endDate`: Filter to date (ISO 8601)
- `limit`: Maximum number of locations (default: 100, max: 10000)

**Response** (200 OK):
```json
{
  "data": [
    {
      "id": "uuid",
      "vehicleId": "uuid",
      "latitude": 40.7128,
      "longitude": -74.0060,
      "accuracy": 5.0,
      "altitude": 10.5,
      "speed": 35.5,
      "heading": 90,
      "type": "GPS",
      "timestamp": "2024-05-04T10:25:00Z"
    }
  ],
  "count": 1000
}
```

#### Record Location
```
POST /api/v1/locations
Content-Type: application/json

{
  "vehicleId": "uuid",
  "latitude": 40.7128,
  "longitude": -74.0060,
  "accuracy": 5.0,
  "speed": 35.5,
  "heading": 90,
  "timestamp": "2024-05-04T10:25:00Z"
}
```
**Response** (201 Created):
```json
{
  "id": "uuid",
  "vehicleId": "uuid",
  "latitude": 40.7128,
  "longitude": -74.0060
}
```

#### Find Nearby Locations
```
GET /api/v1/locations/nearby
?latitude=40.7128&longitude=-74.0060&radiusKm=5&limit=50
```
**Query Parameters**:
- `latitude`: Center point latitude
- `longitude`: Center point longitude
- `radiusKm`: Search radius in kilometers (default: 1, max: 50)
- `limit`: Maximum results (default: 20)

**Response** (200 OK):
```json
{
  "data": [
    {
      "id": "uuid",
      "vehicleId": "uuid",
      "latitude": 40.7180,
      "longitude": -74.0050,
      "distanceKm": 0.8,
      "timestamp": "2024-05-04T10:25:00Z"
    }
  ],
  "centerPoint": {
    "latitude": 40.7128,
    "longitude": -74.0060
  }
}
```

### Route Endpoints

#### List Routes
```
GET /api/v1/routes
?status=Active&vehicleId=uuid&skip=0&take=20
```

#### Create Route
```
POST /api/v1/routes
Content-Type: application/json

{
  "name": "Downtown Delivery",
  "vehicleId": "uuid",
  "waypoints": [
    {
      "order": 1,
      "latitude": 40.7128,
      "longitude": -74.0060,
      "name": "Start Point",
      "estimatedArrivalTime": "2024-05-04T10:30:00Z"
    },
    {
      "order": 2,
      "latitude": 40.7489,
      "longitude": -73.9680,
      "name": "Stop 1",
      "estimatedArrivalTime": "2024-05-04T10:45:00Z"
    }
  ],
  "estimatedDuration": 3600
}
```

## Configuration Guide

### appsettings.json Structure

```json
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=YOUR_SERVER;Database=SignalRMapRealtimeDb;..."
  },
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft.AspNetCore": "Warning"
    }
  },
  "AppSettings": {
    "ApiVersion": "2.0.2",
    "Environment": "Production",
    "EnableSwagger": true,
    "RequestTimeoutSeconds": 30
  },
  "Cors": {
    "AllowedOrigins": ["http://localhost:3000", "http://localhost:5173"]
  },
  "Caching": {
    "Enabled": true,
    "DurationSeconds": 300
  },
  "RateLimiting": {
    "Enabled": true,
    "RequestsPerMinute": 100
  },
  "Notification": {
    "Enabled": true,
    "EmailNotifications": true
  }
}
```

### Environment Variables

```bash
# Database
ASPNETCORE_CONNECTIONSTRINGS_DEFAULTCONNECTION=Server=...

# Application
ASPNETCORE_ENVIRONMENT=Production
ASPNETCORE_URLS=https://0.0.0.0:5001

# Logging
ASPNETCORE_LOGGING_LOGLEVEL_DEFAULT=Information

# SignalR
SIGNALR_BACKPLANETYPE=Redis
SIGNALR_REDIS_CONNECTIONSTRING=localhost:6379
```

### Docker Configuration

See `docker-compose.yml` for complete multi-container setup with SQL Server and Redis.

## Self-Hosting Guide

### Docker Compose (recommended for production)

The included `docker-compose.yml` starts the API, SQL Server, and Redis together.

```bash
# 1. Copy and customise environment variables
cp .env.example .env   # edit DB password, API keys, etc.

# 2. Start all services
docker-compose up -d

# 3. Apply database migrations
docker-compose exec api dotnet ef database update --project SignalRMapRealtime.csproj

# 4. Verify
curl http://localhost:5000/health
```

### Environment Variables

| Variable | Description | Example |
|----------|-------------|---------|
| `ASPNETCORE_ENVIRONMENT` | Runtime environment | `Production` |
| `ASPNETCORE_URLS` | Listening addresses | `http://0.0.0.0:5000` |
| `ConnectionStrings__DefaultConnection` | SQL Server connection string | `Server=db;Database=SignalRMap;User Id=sa;Password=…` |
| `SignalR__Redis__ConnectionString` | Redis backplane (multi-node) | `redis:6379` |

CHANGELOG.md
## [2.0.2] - 2026-05-21
### Fixed
- Fix reconnection loop when server returns 401 during token refresh
- Added regression test for the fix

examples/LocationUpdateSimulator.cs
#nullable enable
// =============================================================================
// Author: Vladyslav Zaiets | https://sarmkadan.com
// CTO & Software Architect
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace SignalRMapRealtime.Examples;

/// <summary>
/// Simulates real-time GPS location updates for multiple vehicles.
/// Useful for testing, development, and load testing.
/// </summary>
public class LocationUpdateSimulator
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;
    private readonly string _apiKey;
    private readonly Random _random = new();

    private readonly List<(double Latitude, double Longitude, string Name)> _routes = new()
    {
        (40.7128, -74.0060, "Times Square"),
        (40.7489, -73.9680, "Midtown"),
        (40.7614, -73.9776, "Central Park"),
        (40.7282, -73.7949, "Queens"),
        (40.6892, -74.0445, "Staten Island")
    };

    public LocationUpdateSimulator(string baseUrl, string apiKey)
    {
        _baseUrl = baseUrl;
        _apiKey = apiKey;
        _httpClient = new HttpClient();
    }

    public async Task RunSimulation(int vehicleCount = 5, int durationSeconds = 60)
    {
        Console.WriteLine($"=== Location Update Simulator ===\n");
        Console.WriteLine($"Simulating {vehicleCount} vehicles for {durationSeconds} seconds...\n");

        var vehicleIds = new List<string>();

        try
        {
            vehicleIds = await CreateVehicles(vehicleCount).ConfigureAwait(false);
            await SimulateMovement(vehicleIds, durationSeconds).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error: {ex.Message}");
        }
    }

    /// <summary>
    /// Creates multiple vehicles in the system.
    /// </summary>
    private async Task<List<string>> CreateVehicles(int count)
    {
        Console.WriteLine($"Creating {count} vehicles...\n");

        var vehicleIds = new List<string>();
        var manufacturers = new[] { "Toyota", "Ford", "BMW", "Mercedes", "Tesla" };
        var models = new[] { "Camry", "Transit", "X5", "E-Class", "Model 3" };

        for (int i = 0; i < count; i++)
        {
            var vehicle = new
            {
                licensePlate = $"DEMO-{i:D3}",
                manufacturer = manufacturers[i % manufacturers.Length],
                model = models[i % models.Length],
                year = 2024,
                vin = Guid.NewGuid().ToString().Replace("-", "").Substring(0, 17)
            };

            var json = JsonSerializer.Serialize(vehicle);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            content.Headers.Add("X-API-Key", _apiKey);

            var response = await _httpClient.PostAsync($"{_baseUrl}/api/v1/vehicles", content).ConfigureAwait(false);

            if (response.IsSuccessStatusCode)
            {
                var result = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                var doc = JsonDocument.Parse(result);
                var vehicleId = doc.RootElement.GetProperty("id").GetString();

                if (vehicleId is not null)
                {
                    vehicleIds.Add(vehicleId);
                    Console.WriteLine($"✓ Created {vehicle.licensePlate}");
                }
            }
        }

        Console.WriteLine();
        return vehicleIds;
    }

    /// <summary>
    /// Simulates vehicle movement by sending periodic location updates.
    /// </summary>
    private async Task SimulateMovement(List<string> vehicleIds, int durationSeconds)
    {
        var startTime = DateTime.UtcNow;
        var endTime = startTime.AddSeconds(durationSeconds);
        var locationCount = 0;

        while (DateTime.UtcNow < endTime)
        {
            foreach (var vehicleId in vehicleIds)
            {
                var location = GenerateLocation(vehicleId);
                var success = await SendLocation(location).ConfigureAwait(false);

                if (success)
                {
                    locationCount++;
                }
            }

            var elapsed = (DateTime.UtcNow - startTime).TotalSeconds;
            var remaining = Math.Max(0, durationSeconds - elapsed);
            Console.WriteLine($"[{elapsed:F0}s] Sent {locationCount} updates (remaining: {remaining:F0}s)");

            await Task.Delay(2000).ConfigureAwait(false); // Update every 2 seconds
        }

        Console.WriteLine($"\n✓ Simulation completed: {locationCount} total location updates sent");
    }

    /// <summary>
    /// Generates a realistic location update with slight variations.
    /// </summary>
    private object GenerateLocation(string vehicleId)
    {
        var route = _routes[_random.Next(_routes.Count)];

        var latitude = route.Latitude + (_random.NextDouble() - 0.5) * 0.01;
        var longitude = route.Longitude + (_random.NextDouble() - 0.5) * 0.01;
        var speed = 20 + _random.Next(50); // 20-70 km/h
        var heading = _random.Next(360);
        var accuracy = 3 + _random.NextDouble() * 7; // 3-10 meters

        return new
        {
            vehicleId = vehicleId,
            latitude = latitude,
            longitude = longitude,
            accuracy = accuracy,
            speed = speed,
            heading = heading,
            type = "GPS",
            timestamp = DateTime.UtcNow
        };
    }

    /// <summary>
    /// Sends a location update to the API.
    /// </summary>
    private async Task<bool> SendLocation(object location)
    {
        try
        {
            var json = JsonSerializer.Serialize(location);
            var content = new StringContent(json, Encoding.UTF8, "application/json");
            content.Headers.Add("X-API-Key", _apiKey);

            var response = await _httpClient.PostAsync($"{_baseUrl}/api/v1/locations", content).ConfigureAwait(false);
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Main entry point for the simulator.
    /// Usage: LocationUpdateSimulator.exe https://localhost:5001 api-key [vehicle-count] [duration-seconds]
    /// </summary>
    public static async Task Main(string[] args)
    {
        var baseUrl = args.Length > 0 ? args[0] : "https://localhost:5001";
        var apiKey = args.Length > 1 ? args[1] : "default-api-key";
        var vehicleCount = args.Length > 2 ? int.Parse(args[2]) : 5;
        var durationSeconds = args.Length > 3 ? int.Parse(args[3]) : 60;

        var simulator = new LocationUpdateSimulator(baseUrl, apiKey);
        await simulator.RunSimulation(vehicleCount, durationSeconds).ConfigureAwait(false);
    }
}
